core/modules/filter/filter.pages.inc

```diff
@@ -505,12 +505,21 @@ function _filter_format_editor_link_url_validate(&$element, &$form_state) {
     }
   }
 
-  form_set_value($element, $value, $form_state);
-
   // Unlike #type = 'url' validation, we allow both relative and absolute paths.
   if ($value !== '' && !valid_url($value, TRUE) && !valid_url($value, FALSE)) {
-    form_error($element, t('The URL %url is not valid.', array('%url' => $value)));
+    // We might need to encode some characters, but we only do this if
+    // transliteration is off and only for relative paths.
+    $encoded_value = backdrop_encode_path($value);
+    if (config_get('path.settings', 'transliterate') == FALSE && valid_url($encoded_value, FALSE)) {
+      // Function url() takes care of base_path and language prefix.
+      $value = url($value);
```

Comment:
@indigoxela I haven't actually tested this theory, but looking at the code/logic, I think that the outer-most `if` in this chunk of code should only check for `$value !== ''`. Then we should be doing something like `$valid_url = valid_url($value, TRUE) || valid_url($value, FALSE);`.

Also: https://docs.backdropcms.org/api/backdrop/1/search/backdrop_encode_path

> Note that url() takes care of calling this function, so a path passed to that function should not be encoded in advance.

In general, perhaps we should be doing something like this:

```php
$value = url($value);

// Check whether we have a valid URL (after having potentially encoded).
// Unlike #type = 'url' validation, we allow both relative and absolute paths.
$valid_url = valid_url(value, TRUE) || valid_url(value, FALSE);

if ($valid_url) {
  // Do things...
}
else {
  form_error($element, t('The URL %url is not valid.', array('%url' => $value)));
}
```

Reply:
I see, interesting points.

Moving the check for empty value further to top absolutely makes sense :+1: 

Using url() as a swiss army knife here could potentially enable us to get rid of the (hidden) link module dependency (another issue) in _filter_format_editor_link_url_validate() - but that needs a closer look, as it might have side effects.

Blindly adding the leading slash for internal links - when the link module is available - has glitches anyway (yet another issue).

But that would expand the scope of this PR beyond the linked issue... And I suspect, we'd need some functional tests. I'll think it over (too busy ATM).

Reply:
> Moving the check for empty value further to top absolutely makes sense 👍

Yes. I realized that even my implementation suggestion seems not 100% 🤔 ...I'm thinking that ideally, we should be "breaking early" it the value is blank, and consider the entered URL invalid ...no need to even call `link_validate_url()` or `link_cleanup_url()` in that case ...not even point for `module_exists('link')`.

>I'll think it over (too busy ATM).

Thanks for all the time/energy you spend in making Backdrop better 🙏 ❤️

Reply:
I ended up doing a complete rewrite - for several reasons. I'll explain in the issue thread.